### ISSUE TITLE
Consider a class-based API

### ISSUE THREAD

**Issue by tobyzerner** [tobyzerner/json-api-server#37]
It could look like:

```php
class Posts extends ResourceType
{
    public static $type = 'posts';

    public function adapter(): AdapterInterface
    {
        return new EloquentAdapter(QuestionRevision::class);
    }

    public function scope(Builder $query, Context $context): void
    {
        $query->where('user_id', $context->getRequest()->getAttribute('userId'));
    }

    public function endpoints(): array
    {
        return [
            Endpoint::show(),
            Endpoint::create()->authorized(fn($context) => true),
            Endpoint::update(),
        ];
    }

    public function fields(): array
    {
        return [
            String::make('title')->writable(),
            String::make('body')->writable(),
            Boolean::make('isPublished')->writable(),
        ];
    }
}
```

Please comment below if you have any thoughts about this vs. the current API.

**Comment by tobyzerner** [comment]
@franzliedke @smtlab any thoughts on this?

**Comment by smtlab** [comment]
This is good, but the current closure-based API is flexible and can be hooked into any existing MVC controller, Service or framework. Class-based API would be a good dedicated implementation specific to this package.

**Comment by franzliedke** [comment]
I like it, but also think it should be optional.

**Comment by tobyzerner** [comment]
New class-based API released in [v1.0.0-alpha.1](https://github.com/tobyzerner/json-api-server/releases/tag/v1.0.0-alpha.1)! Sorry, I dumped the old API because the new one is vastly superior :)
